Questions with deeply nested MBQL can be exported, but not imported using remote sync

### Describe the bug

We fixed https://github.com/metabase/metabase/issues/71257 by failing a remote sync import rather than partially succeeding if any resource could not be imported. The most important issue of partial syncs is solved, but customers that run into this situation are still stuck and cannot use remote sync until they manually fix their content.

### To Reproduce

1. Create a deeply nested query per steps in https://github.com/metabase/metabase/issues/71257
2. Try to import it using remote sync

<img width="1512" height="944" alt="Image" src="https://github.com/user-attachments/assets/9010e929-ff63-4a58-9c8b-561a4ab84ac0" />

### Expected behavior

If we've allowed customers to save something and export it, it should be possible to import that as well. Metabase should never generate invalid exports.

If we don't want to consider such deeply nested MBQL queries as valid, then customers should not be able to save or export those queries.

Another quick solution could be to expose the max YAML nesting as a setting, and for customers to change that as needed. So far this has only shown up once that we've heard of.

### Logs

<details><summary>Details</summary>
<p>

```
2026-03-27 15:19:51,772 WARN source.ingestable :: Error parsing file during ingestion: Nesting Depth exceeded max 50 
2026-03-27 15:19:51,895 ERROR remote-sync.impl :: Failed to reload from git repository: Failed to read 1 file(s) during ingestion: collections/92QELeRLYTkA04idC7LbA_my_important_collection/cards/T1Dzz_PhhBBkeCdDBApXK_in2deep.yaml 
clojure.lang.ExceptionInfo: Failed to read 1 file(s) during ingestion: collections/92QELeRLYTkA04idC7LbA_my_important_collection/cards/T1Dzz_PhhBBkeCdDBApXK_in2deep.yaml
	at metabase_enterprise.serialization.v2.load$load_metabase_BANG_$with_connection_STAR___144675$with_transaction_STAR___144676.invoke(load.clj:225)
	at toucan2.connection$bind_current_connectable_fn$fn__34264.invoke(connection.clj:104)
	at metabase.app_db.connection$do_transaction$thunk__41392.invoke(connection.clj:146)
	at metabase.app_db.connection$do_transaction.invokeStatic(connection.clj:165)
	at metabase.app_db.connection$do_transaction.invoke(connection.clj:142)
	at metabase.app_db.connection$do_with_transaction_primary_method_java_sql_Connection.invokeStatic(connection.clj:206)
	at metabase.app_db.connection$do_with_transaction_primary_method_java_sql_Connection.invoke(connection.clj:179)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invokeStatic(connection.clj:249)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invoke(connection.clj:245)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:58)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:212)
	at metabase_enterprise.serialization.v2.load$load_metabase_BANG_$with_connection_STAR___144675.invoke(load.clj:214)
	at toucan2.connection$bind_current_connectable_fn$fn__34264.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__34264.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__34264.invoke(connection.clj:104)
	at toucan2.jdbc.connection$do_with_connection_primary_method_javax_sql_DataSource.invokeStatic(connection.clj:18)
	at toucan2.jdbc.connection$do_with_connection_primary_method_javax_sql_DataSource.invoke(connection.clj:15)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at metabase.app_db.connection$do_with_connection_primary_method_default.invokeStatic(connection.clj:133)
	at metabase.app_db.connection$do_with_connection_primary_method_default.invoke(connection.clj:131)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at toucan2.connection$do_with_connection_primary_method_.invokeStatic(connection.clj:204)
	at toucan2.connection$do_with_connection_primary_method_.invoke(connection.clj:194)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__32107.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at metabase_enterprise.serialization.v2.load$load_metabase_BANG_.invokeStatic(load.clj:214)
	at metabase_enterprise.serialization.v2.load$load_metabase_BANG_.doInvoke(load.clj:206)
	at clojure.lang.RestFn.invoke(RestFn.java:413)
	at metabase_enterprise.remote_sync.impl$import_BANG_$fn__146286.invoke(impl.clj:269)
	at metabase_enterprise.remote_sync.impl$import_BANG_.invokeStatic(impl.clj:268)
	at metabase_enterprise.remote_sync.impl$import_BANG_.doInvoke(impl.clj:225)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase_enterprise.remote_sync.impl$async_import_BANG_$fn__146398.invoke(impl.clj:479)
	at metabase_enterprise.remote_sync.impl$run_async_BANG_$fn$reify__146378$fn__146379.invoke(impl.clj:455)
	at metabase_enterprise.remote_sync.impl$run_async_BANG_$fn$reify__146378.get(impl.clj:454)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.TimeoutExecutor.lambda$apply$1(TimeoutExecutor.java:93)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at metabase_enterprise.remote_sync.impl$run_async_BANG_$fn__146377.invoke(impl.clj:451)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1990)
	at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1990)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:135)
	at clojure.core$apply.invokeStatic(core.clj:671)
	at clojure.core$bound_fn_STAR_$fn__5839.doInvoke(core.clj:2020)
	at clojure.lang.RestFn.invoke(RestFn.java:400)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.lang.VirtualThread.run(Unknown Source)
Caused by: clojure.lang.ExceptionInfo: Failed to parse file: collections/92QELeRLYTkA04idC7LbA_my_important_collection/cards/T1Dzz_PhhBBkeCdDBApXK_in2deep.yaml
	at metabase_enterprise.remote_sync.source.ingestable$ingest_all$iter__145331__145335$fn__145336$fn__145373.invoke(ingestable.clj:39)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_all$iter__145331__145335$fn__145336.invoke(ingestable.clj:34)
	at clojure.lang.LazySeq.force(LazySeq.java:50)
	at clojure.lang.LazySeq.realize(LazySeq.java:89)
	at clojure.lang.LazySeq.seq(LazySeq.java:106)
	at clojure.lang.Cons.next(Cons.java:41)
	at clojure.lang.RT.next(RT.java:733)
	at clojure.core$next__5472.invokeStatic(core.clj:64)
	at clojure.core.protocols$fn__8279.invokeStatic(protocols.clj:168)
	at clojure.core.protocols$fn__8279.invoke(protocols.clj:123)
	at clojure.core.protocols$fn__8233$G__8228__8242.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
	at clojure.core.protocols$fn__8266.invokeStatic(protocols.clj:74)
	at clojure.core.protocols$fn__8266.invoke(protocols.clj:74)
	at clojure.core.protocols$fn__8207$G__8202__8220.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6965)
	at clojure.core$into.invokeStatic(core.clj:7038)
	at clojure.core$into.invoke(core.clj:7029)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_all.invokeStatic(ingestable.clj:25)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_all.invoke(ingestable.clj:19)
	at metabase_enterprise.remote_sync.source.ingestable.IngestableSnapshot$fn__145470.invoke(ingestable.clj:124)
	at metabase_enterprise.remote_sync.source.ingestable$populate_cache_BANG_.invokeStatic(ingestable.clj:116)
	at metabase_enterprise.remote_sync.source.ingestable$populate_cache_BANG_.invoke(ingestable.clj:114)
	at metabase_enterprise.remote_sync.source.ingestable.IngestableSnapshot.ingest_list(ingestable.clj:124)
	at metabase_enterprise.remote_sync.impl$snapshot_has_transforms_QMARK_.invokeStatic(impl.clj:35)
	at metabase_enterprise.remote_sync.impl$snapshot_has_transforms_QMARK_.invoke(impl.clj:26)
	at metabase_enterprise.remote_sync.impl$import_BANG_.invokeStatic(impl.clj:248)
	... 27 more
Caused by: org.yaml.snakeyaml.error.YAMLException: Nesting Depth exceeded max 50
	at org.yaml.snakeyaml.composer.Composer.increaseNestingDepth(Composer.java:404)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:211)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:284)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:216)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:396)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:361)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:329)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:218)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:396)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:361)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:329)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:218)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:396)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:361)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:329)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:218)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:396)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:361)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:329)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:218)
	at org.yaml.snakeyaml.composer.Composer.getNode(Composer.java:141)
	at org.yaml.snakeyaml.composer.Composer.getSingleNode(Composer.java:167)
	at org.yaml.snakeyaml.constructor.BaseConstructor.getSingleData(BaseConstructor.java:179)
	at org.yaml.snakeyaml.Yaml.loadFromReader(Yaml.java:508)
	at org.yaml.snakeyaml.Yaml.load(Yaml.java:461)
	at clj_yaml.core$load_stream.invokeStatic(core.clj:279)
	at clj_yaml.core$load_stream.invoke(core.clj:275)
	at clj_yaml.core$parse_string.invokeStatic(core.clj:325)
	at clj_yaml.core$parse_string.doInvoke(core.clj:281)
	at clojure.lang.RestFn.invoke(RestFn.java:426)
	at metabase.util.yaml$parse_string.invokeStatic(yaml.clj:52)
	at metabase.util.yaml$parse_string.doInvoke(yaml.clj:49)
	at clojure.lang.RestFn.invoke(RestFn.java:426)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_content.invokeStatic(ingestable.clj:17)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_content.invoke(ingestable.clj:15)
	at metabase_enterprise.remote_sync.source.ingestable$ingest_all$iter__145331__145335$fn__145336$fn__145373.invoke(ingestable.clj:36)
	... 53 more
```

</p>
</details> 

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:148.0) Gecko/20100101 Firefox/148.0",
    "vendor": ""
  },
  "metabase-info": {
    "valid-thru": "2026-12-31T12:00:00Z",
    "current-user-count": 2,
    "application-database": "postgres",
    "settings": {
      "report-timezone": null
    },
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "17.5 (Debian 17.5-1.pgdg120+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.8"
      }
    },
    "databases": [
      "postgres",
      "h2"
    ],
    "max-users": 50,
    "plan-alias": "enterprise-airgap",
    "hosting-env": "unknown",
    "run-mode": "prod",
    "version": {
      "date": "2026-03-26",
      "tag": "v1.60.0.1-beta",
      "hash": "eed1d3a"
    },
    "airgap-token": "enabled"
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.10+7-LTS",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "21.0.10",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.10+7-LTS",
    "os.name": "Linux",
    "os.version": "6.17.8-orbstack-00308-g8f9c941121b1",
    "user.language": "en",
    "user.timezone": "America/Argentina/Buenos_Aires"
  }
}
```

### Severity

P2

### Additional context

_No response_

--- a/src/metabase/util/yaml.clj
+++ b/src/metabase/util/yaml.clj
@@ -9,17 +9,35 @@
    [metabase.util :as u]
    [metabase.util.date-2 :as u.date]
    [metabase.util.files :as u.files]
-   [metabase.util.log :as log]
-   [potemkin :as p])
+   [metabase.util.log :as log])
   (:import
    (java.nio.file Files Path)
    (java.time.temporal Temporal)))
 
 (set! *warn-on-reflection* true)
 
-(p/import-vars
- [yaml
-  parse-stream])
+(def ^:const nesting-depth-limit
+  "Maximum YAML nesting depth allowed when parsing.
+
+  SnakeYAML defaults to 50, which is too shallow for deeply nested MBQL. MBQL sometimes contains expressions with many
+  nested function calls like `replace(replace(...))`. There's no depth limit on the YAML output, so they are perfectly
+  valid queries that can be saved, executed and exported. But with the default depth limit, they cannot be imported,
+  leaving remote-sync users stuck (see #71257 / UXW-3770).
+
+  We raise the limit so that anything Metabase can export can also be re-imported, while keeping a finite bound as a
+  sanity guard against pathological/malicious input. Callers may override via the `:nesting-depth-limit` opt."
+  10000)
+
+(defn- with-parse-defaults
+  "Merges Metabase's default parsing options into `opts`; preserving any values the caller supplied."
+  [opts]
+  (merge {:nesting-depth-limit nesting-depth-limit} opts))
+
+(defn parse-stream
+  "Returns Clojure data structures for a stream of YAML read from `reader`, with opts passed to
+  clj-yaml. Applies Metabase's default parsing options (see [[with-parse-defaults]])."
+  [reader & {:as opts}]
+  (yaml/parse-stream reader (with-parse-defaults opts)))
 
 (defn- vectorized
   "Returns x with lazy seqs converted to vectors wherever they appear in the data structure."
@@ -47,9 +65,10 @@
   (yaml/generate-string x opts))
 
 (defn parse-string
-  "Returns a Clojure object parsed from YAML in string s with opts passed to clj-yaml."
+  "Returns a Clojure object parsed from YAML in string s with opts passed to clj-yaml.
+  Applies Metabase's default parsing options (see [[with-parse-defaults]])."
   [s & {:as opts}]
-  (vectorized (yaml/parse-string s opts)))
+  (vectorized (yaml/parse-string s (with-parse-defaults opts))))
 
 ;; Legacy API:
 

--- a/test/metabase/util/yaml_test.clj
+++ b/test/metabase/util/yaml_test.clj
@@ -1,0 +1,46 @@
+(ns metabase.util.yaml-test
+  (:require
+   [clojure.test :refer :all]
+   [metabase.util.yaml :as yaml])
+  (:import
+   (java.io File StringReader)))
+
+(set! *warn-on-reflection* true)
+
+(defn- nested-structure
+  "Builds a value nested `depth` levels deep, e.g. `(nested-structure 3)` => `[[[\"leaf\"]]]`.
+
+  This mimics the deeply nested YAML produced by serializing MBQL expressions with many nested
+  function calls (see #71257 / UXW-3770)."
+  [depth]
+  (-> (iterate vector "leaf")
+      (nth depth)))
+
+(deftest ^:parallel parse-string-allows-deep-nesting-test
+  (testing "parse-string allows nesting deeper than SnakeYAML's default limit of 50 (#71257)"
+    (let [deep (nested-structure 500)]
+      (is (= deep
+             (-> deep yaml/generate-string yaml/parse-string))
+          "deeply nested YAML that Metabase can generate must round-trip back through parse-string")))
+  (testing "callers may still tighten the nesting limit via :nesting-depth-limit"
+    (is (thrown-with-msg? Exception #"Nesting Depth exceeded"
+                          (-> (nested-structure 500)
+                              yaml/generate-string
+                              (yaml/parse-string :nesting-depth-limit 10))))))
+
+(deftest ^:parallel from-file-allows-deep-nesting-test
+  (testing "from-file allows nesting deeper than SnakeYAML's default limit (this is the serdes import path)"
+    (let [deep      (nested-structure 500)
+          ^File tmp (File/createTempFile "deep-yaml" ".yaml")]
+      (try
+        (spit tmp (yaml/generate-string deep))
+        (is (= deep (yaml/from-file tmp)))
+        (finally
+          (.delete tmp))))))
+
+(deftest ^:parallel parse-stream-allows-deep-nesting-test
+  (testing "parse-stream allows nesting deeper than SnakeYAML's default limit"
+    (let [deep (nested-structure 500)
+          s    (yaml/generate-string deep)]
+      (with-open [r (StringReader. s)]
+        (is (= deep (yaml/parse-stream r)))))))

